src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectState.cs

```diff
@@ -152,17 +163,36 @@ public ProjectState WithAddedHostDocument(HostDocument hostDocument, Func<Task<T
             {
                 return this;
             }
+            
+            var documents = Documents.Add(hostDocument.FilePath, DocumentState.Create(Services, hostDocument, loader));
 
-            var documents = new Dictionary<string, DocumentState>(FilePathComparer.Instance);
-            foreach (var kvp in Documents)
+            // Compute the effect on the import map
+            ImmutableArray<string> relatedDocuments;
+
+            var importsToRelatedDocuments = ImportsToRelatedDocuments;
+            var importTargetPaths = ProjectEngine.GetImportDocumentTargetPaths(this, hostDocument.TargetPath);
+            foreach (var importTargetPath in importTargetPaths)
             {
-                documents.Add(kvp.Key, kvp.Value);
+                if (!importsToRelatedDocuments.TryGetValue(importTargetPath, out relatedDocuments))
+                {
+                    relatedDocuments = ImmutableArray.Create<string>();
+                }
+
+                relatedDocuments = relatedDocuments.Add(hostDocument.FilePath);
+                importsToRelatedDocuments = importsToRelatedDocuments.SetItem(importTargetPath, relatedDocuments);
```

Comment:
Just to be clear, with how this is currently it doesn't account for any documents that were removed from the underlying ProjectFileSystem that haven't yet been processed.  Aka, if Document A existed in `importsToRelatedDocuments` before calling `WithAddedHostDocument` and then various changes occurred simultaneously (like a user moving an entire folder) then this would have a load of documents in it that don't exist on disk.

Maybe that's ok, maybe it's not. Bringing it up.

Reply:
I don't understand this. if the files are deleted, their content would change and that would trigger a notification.

Reply:
`ProjectFileSystem` interacts with the file system directly so while the system is waiting for events to propagate there may be a few intermediary project states which aren't entirely correct
.

Reply:
We can't and shouldn't control the ordering of events - we can only work to make sure that the result is eventually coherent which seems like it will be the case.